prometheus/vec_test.go

```diff
@@ -125,6 +125,87 @@ func testDeleteLabelValues(t *testing.T, vec *GaugeVec) {
 	}
 }
 
+func TestDeletePartialMatch(t *testing.T) {
+	vec := NewGaugeVec(
+		GaugeOpts{
+			Name: "test",
+			Help: "helpless",
+		},
+		[]string{"l1", "l2"},
+	)
+	testDeletePartialMatch(t, vec)
+}
+
+func testDeletePartialMatch(t *testing.T, vec *GaugeVec) {
+	// No metric value is set.
+	if got, want := vec.DeletePartialMatch(Labels{"l1": "v1", "l2": "v2"}), 0; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+
+	c1 := vec.MustCurryWith(Labels{"l1": "v1"})
+	c1.WithLabelValues("2").Inc()
+
+	// Try to delete nonexistent label lx with existent value v1.
+	if got, want := c1.DeletePartialMatch(Labels{"lx": "v1"}), 0; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+
+	// Delete with valid pair l1: v1.
+	if got, want := c1.DeletePartialMatch(Labels{"l1": "v1"}), 1; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+
+	// Try to delete with partially invalid labels.
+	vec.With(Labels{"l1": "v1", "l2": "v2"}).(Gauge).Set(42)
+	if got, want := vec.DeletePartialMatch(Labels{"l1": "v1", "l2": "xv2"}), 0; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+
+	// Try to delete with a single valid label which matches multiple metrics.
+	vec.With(Labels{"l1": "v1", "l2": "v2"}).(Gauge).Set(42)
+	vec.With(Labels{"l1": "v1", "l2": "vv22"}).(Gauge).Set(84)
+	c3 := vec.MustCurryWith(Labels{"l2": "l2C3CurriedValue"}) // Used below
+	vec.With(Labels{"l1": "v3", "l2": "v3"}).(Gauge).Set(168)
+	if got, want := vec.DeletePartialMatch(Labels{"l1": "v1"}), 2; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
+
+	// Try to delete a value which shouldn't be in our base vector (only the curried one c3).
+	if got, want := vec.DeletePartialMatch(Labels{"l2": "l2C3CurriedValue"}), 0; got != want {
+		t.Errorf("got %v, want %v", got, want)
+	}
```

Comment:
Note that this doesn't delete anything because there is no matching element neither in the base vector nor in the curried vector (which is only a "view" on the base vector). Currying alone doesn't create an element in the vector. You needed to call something like `c3.With(Labels{"l1": "something"})` first. After that, the test as written here (`vec.DeletePartialMatch(Labels{"l2": "l2C3CurriedValue"})`) _would_ actually delete an element.